### DISCUSSION TITLE
Hard sector support status?

### DISCUSSION THREAD

**Discussion by nullvalue0** [keirf/flashfloppy#735] [Q&A]
I see there have been various attempts to add hard sector support to FlashFloppy. Looks like issues 393, 400 & 412.

In particular I have a Micropolis/Vector Graphic S-100 computer that I'd love to be able to use with a Gotek (VGI images available here: https://deramp.com/downloads/vector_graphic/software/disk_images/Micropolis%20controller/)

Just wondering what the current status is?

**Comment by keirf** [comment]
There's beta support in latest experimental v4 firmware. You will need a suitable hard-sectored HFEv3 image file. There is a script in #400 which is tailored for Micropolis. And also you can ask @ejona86 about it.

**Reply by ejona86** [reply]
I have a Vector 4, and the hard sector support is solid on it. It uses a different controller board than earlier Vectors, which it sounds like you have. We had trouble on some other hard-sectored machines.

https://github.com/keirf/flashfloppy/issues/400#issuecomment-898975665 is the part of #400 to look at. The config is still applicable and `scripts/mk_hfe.py` can make you a blank disk with the `--hard-sectors 16` flag. Generally you should use the latest v4 release, which is v4.7a right now.

You'll be the first to find this script convenient:
```python
# vgi2hfe.py
#
# Convert VGI to HFE image.
#
# Written & released by Keir Fraser <keir.xen@gmail.com> and Eric Anderson
# <ejona86@gmail.com>
#
# This is free and unencumbered software released into the public domain.
# See the file COPYING for more details, or visit <http://unlicense.org>.

import sys,struct,argparse

def mfm(bs):
  os = bytearray()
  last_d = False
  o = 0
  for b in bs:
    for _ in range(2):
      for _ in range(4):
        o >>= 2
        d = bool(b & 0x80)
        o |= (not (d or last_d)) << 6
        o |= d << 7
        last_d = d
        b <<= 1
      os.append(o)
  return os

def main(argv):
  parser = argparse.ArgumentParser(
    formatter_class=argparse.ArgumentDefaultsHelpFormatter)
  parser.add_argument("--rate", type=int, default=250,
                      help="data rate, kbit/s")
  parser.add_argument("--rpm", type=int, default=300,
                      help="rotational rate, rpm")
  parser.add_argument("--cyls", type=int, default=77,
                      help="number of cylinders")
  parser.add_argument("--sides", type=int, default=2,
                      help="number of sides")
  parser.add_argument("--hard-sectors", type=int, default=16,
                      help="number of hard sectors")
  parser.add_argument("--infile-sector-size", type=int, default=275,
                      help="bytes per sector")
  parser.add_argument("--strip-ecc", type=bool, default=False,
                      help="Strip ECC")
  parser.add_argument("infile", help="input filename")
  parser.add_argument("outfile", help="output filename")
  args = parser.parse_args(argv[1:])

  v3 = args.hard_sectors != 0

  bits = (args.rate * 1000 * 60) // args.rpm
  bits *= 2 # clock bits
  bits *= 2 # 2 sides
  byte_cnt = (bits + 7) // 8 # convert to bytes, rounded up
  if not v3:
    byte_cnt = (byte_cnt + 15) & ~15 # round up to 16-byte boundary
  raw_bytes = byte_cnt
  if args.hard_sectors:
    byte_cnt += (args.hard_sectors + 1) * 2 # 2 sides
  blocks = (byte_cnt + 511) // 512 # convert to 512-byte blocks, rounded up

  print("Geometry: %u cylinders, %u sides" % (args.cyls, args.sides))
  print("%ukbit/s @ %uRPM -> %u Encoded Bits"
        % (args.rate, args.rpm, bits/2))
  print("Data per HFE Track: %u bytes, %u blocks" % (byte_cnt, blocks))

  in_f = open(args.infile, "rb")
  
  # Header
  out_f = open(args.outfile, "wb")
  sig = b'HXCHFEV3' if v3 else b"HXCPICFE"
  out_f.write(struct.pack("<8s4B2H2BH",
                          sig,        # signature
                          0,          # revision
                          args.cyls,  # nr_tracks
                          args.sides, # nr_sides
                          0xff,       # track_encoding
                          args.rate,  # bitrate
                          0,          # rpm
                          0xfe,       # interface_mode
                          1,          # rsvd
                          1))         # track_list_offset
  out_f.write(bytearray(b'\xff'*(512-20)))

  # TLUT
  tlut_blocks = (args.cyls*4 + 511) // 512
  base = 1 + tlut_blocks
  for i in range(args.cyls):
    out_f.write(struct.pack("<2H", base, byte_cnt))
    base += blocks
  out_f.write(bytearray(b'\xff'*(tlut_blocks*512-args.cyls*4)))

  # Data
  if not args.hard_sectors:
    out_f.write(bytearray(b'\x88'*(blocks*512*args.cyls)))
  else:
    strip_bytes = 5 if args.strip_ecc else 0
    for cyl in range(args.cyls):
      side = []
      for ss in range(2):
        trk_side = bytearray(b'\x88')*(raw_bytes//2)
        sector_len = float(raw_bytes//2) / args.hard_sectors
        if args.sides == 2 or ss == 0:
          in_base = cyl * args.infile_sector_size * args.hard_sectors
          if ss == 1:
            in_base += args.cyls * args.infile_sector_size * args.hard_sectors
          last_sector_pos = len(trk_side)
          for sector in range(args.hard_sectors-1, -1, -1):
            in_f.seek(in_base + sector * args.infile_sector_size)
            raw_b = bytes(40) + in_f.read(args.infile_sector_size-strip_bytes) + bytes(1)
            sector_pos = int(sector_len*sector)
            raw_mfm = mfm(raw_b)
            raw_mfm += b'\x55' * (last_sector_pos-sector_pos-len(raw_mfm))
            trk_side[sector_pos:sector_pos+len(raw_mfm)] = raw_mfm
            last_sector_pos = sector_pos
        trk_side.insert(-int(sector_len)//2, 0x8F)
        for sector in range(args.hard_sectors-1, -1, -1):
          trk_side.insert(int(sector_len*sector), 0x8F)
        assert len(trk_side) == byte_cnt//2
        trk_side += b'\x0F'*(256 - len(trk_side)%256)
        assert len(trk_side) == blocks*256
        side.append(trk_side)
      if args.sides == 1:
        side.append(trk_side)

      trk = bytearray()
      for pos in range(0, len(side[0]), 256):
        trk.extend(side[0][pos:pos+256])
        trk.extend(side[1][pos:pos+256])
      out_f.write(trk)
    
if __name__ == "__main__":
    main(sys.argv)
```

**Reply by nullvalue0** [reply]
Thanks again & sorry about the delay - was waiting for a new Gotek & a couple parts to come in. Wanted to do this right & build a dedicated 5V power supply to the Gotek drive from the +8VDC line. Using a nice switching 5V DC/DC regulator as opposed to just a 7805. Got everything all wired up and ready to go, but no luck yet.

What I've done so far..

1. Applied 4.7a update to Gotek
2. Used your linked python script to covert some of the .VGI images to .HFE. This process seemed to work well. Going through them, I think the one I need is VG09 as this is labeled for a 56k system which mine is.
3. Copied the images to a "Vector Graphic" folder on the USB
4. Applied the 3 FF.CFG settings in your comment (however for index-suppression, you put false but it expects yes|no. I tried both, no difference)
5. Boot the VG system to Monitor 4.0. RAM completes a check of all 56k.
6. Tried both the 'B' boot option, and "G F800" - neither seem to do anything. Nothing happens on the Gotek (track seeks etc), and the system just hangs.

So I'm not really sure where to start debugging, what would you recommend?

A few action shots...
![IMG_20230111_184939](https://user-images.githubusercontent.com/58637612/211951559-4a848569-f131-4307-b181-481f6d8733ed.jpg)
![IMG_20230111_185035](https://user-images.githubusercontent.com/58637612/211951553-1789c4dd-0cad-4fe1-9648-7ef88b565390.jpg)
![IMG_20230111_185018](https://user-images.githubusercontent.com/58637612/211951550-8656031d-e3bb-410e-ad85-c7d0e2b8d9bc.jpg)
![IMG_20230111_185002](https://user-images.githubusercontent.com/58637612/211951548-1558f570-5f65-4327-8064-2dfcf4f61684.jpg)

**Reply by ejona86** [reply]
@nullvalue0, if it just hangs without any changes on the screen, then that normally meant for me that I needed to adjust the jumpers. It should really at least try to seek. From what I can tell, you should just have a jumper on S0. Remove the other gotek jumpers. If nothing happens, move the jumper to S1, and if still nothing move the jumper to M0.

Do you have any working real floppy drive and floppies? I looked at VG09, and it seems like a good bet. Looks like reasonable boot sector code and "VECTOR GRAPHIC 56K CP/M - VERSION 2.20".

I started reading through https://bitsavers.org/pdf/vectorGraphic/hardware/Micropolis_Disk_Controller_Board.pdf just to find the boot disk number and the troubleshooting section was surprisingly unhelpful; I guess the "User's Information Sheet" (which I can't find) may have been better. It mentions a bootstrap listing in the "Micropolis User Guide" which I think is http://www.bitsavers.org/pdf/micropolis/metafloppy/1084-01_1040_1050_Users_Manual_Apr79.pdf 
 . The bootstrap in Appendix F, page F-6, appears to use disk 0, which is why I suggested S0 above.

**Reply by nullvalue0** [reply]
@ejona86 Well, that's a little progress. Setting the jumper to S0 now when I attempt to boot, the gotek LED comes on, I see the tracks very quickly counts up from 0 to about 7, then back down to 0. Then the LED stays on but nothing else happens & nothings shows up on the computer. Jumper was originally in S1 position. Tried M0 and S1 again and nothing happens there - so I think you're right with the S0 position.

I got no disks with the system. I do have a Micropolis 1053-II dual drive, however it is not operational.

**Reply by hharte** [reply]
@nullvalue0 It looks like you have a FlashWriter II video card.  I’m not sure if the boot disk you have is for the FW2 or a serial console.  One idea is to try that .vgi with the SIMH/AltairZ80 simulator and see whether it uses the FW2 or serial port for console:
https://schorn.ch/altair_5.php

7 tracks * 16 * 256 = 28K, so it’s possible CP/M booted but it is not using the monitor ROM routines for console I/O.

There is also a document describing the contents of many of the .vgi images here: https://deramp.com/downloads/vector_graphic/software/disk_images/Micropolis%20controller/56K%20Disk%20Directory.pdf

**Reply by nullvalue0** [reply]
@hharte That's a really good thought. I haven't had a chance to try it in the simulator - but I did hook up a serial cable. I am able to verify serial communications are working through the monitor's commands I and O. However, when attempting to boot, I don't get anything across the line. My Bit Streamer is addressed at port 6/7 - not sure if maybe the disks are set to address a different port.

Update: I did have chance to try out the simulator with the VG09 disk image, it seems to load just fine.. But I'm not sure how to tell if this is being simulated through a serial port or FlashWriter? How can I tell the difference? Here's the sim output...not sure why I'm getting some funky characters, this is under Windows 10:

`                    --VECTOR GRAPHIC MONITOR VERSION 4.0 C--
ì                                COPYRIGHT  1979


è
Mon>áMEM CHEC╦
E000 FF C3
Mon>áBOOT DIS╦
VECTOR GRAPHIC 56K CP/M - VERSION 2.20

A>dir
A: MOVCPM   COM | SYSGEN   COM | SUBMIT   COM | XSUB     COM
A: MBASIC4  COM | MBASIC5  COM | ED       COM | SC       COM
A: ZSM      COM | RAID     COM | DDT      COM | LOAD     COM
A: DUMP     COM | STAT     COM | PIP      COM | FORMAT   COM
A: BACKUP   COM | CONFIG   COM | MAP      COM | WORM     COM
A: FLASH10  COM | MDIAG    COM | USERAREA ASM | USERCUST ASM
A: DUMP     ASM | DUMP     HEX | DUMP     PRN | XFER     COM
A>`

**Reply by nullvalue0** [reply]
Still playing around just trying to see what I can do.. I wanted to see how far I could get with the PC2FLOP process (https://deramp.com/downloads/vector_graphic/software/disk_image_transfer/Micropolis%20controller/)

I used your "mk_hfe.py --hard-sectors 16" script to generate a few blank images to play with.. Got the PC2FLOP.COM program transferred to the computer and running. I have a little prior experience with this app from when I was trying to get my physical drive working.. It works by sending a VGI disk image over XMODEM - it buffers the data up in RAM and writes it out to disk in chunks. I actually made it all the way through the process!  The Gotek LED would come on, 'W' would show up, and tracks slowly advanced all the way to the 70s (76 probably? wasn't paying attention) - and got a "successful" message at the end.

Reviewing the disk image that got written to USB back over on the PC, it definitely wrote stuff. Whether it's right, I have no idea. Unfortunately I wasn't able to boot it - it does the same thing as the other images - tracks quickly count up to 7 and back to 0, then hangs..

Anyways, thought I'd share - Writes seem to be working from what I can test. Just can't boot yet..  I suppose I should give the FLOP2PC process a shot just to see what happens?

![IMG_20230112_210345~2](https://user-images.githubusercontent.com/58637612/212229056-fa6c488c-567f-4d1c-b1db-4bf1ed5ce5bd.jpg)
![IMG_20230112_210810~3](https://user-images.githubusercontent.com/58637612/212229075-927e44de-4558-4806-acf3-f490952e26f0.jpg)
![IMG_20230112_211248~3](https://user-images.githubusercontent.com/58637612/212229087-03b13039-87e1-43dc-a65f-a0eb50e7f5de.jpg)

**Reply by hharte** [reply]
@nullvalue0 I tried VG09 with both serial and FWII in SIMH, and both work, so that's not the issue.

Here is the SIMH configuration file I used.  Monitor 4.0 (seems to be the same one your real machine is using) is available here: https://deramp.com/downloads/vector_graphic/software/roms/Version%204.x%20Monitors/

It is an Intel HEX file, but can be converted to binary and loaded into the simulator.  You can experiment with the config file below, uncommenting the monitor for either serial or FWII.

```
; Startup file for the Vector Graphics simulator.
; Simulating a 56K Vector MZ System, with Micropolis FD Controller
; and Flashwriter
set cpu 64k
set cpu nonbanked
set cpu itrap
set cpu z80
set cpu noaltairrom
set sio del
set mdsk enabled
set fwii enabled

; Load FlashWriter II Monitor 4.0
load mon40.bin e000 rom

; Load serial monitor
;load mon40c.bin e000 rom
att mdsk0 vg09.vgi

; Flashwriter2 at 0xF000
att fwii0 f000

g e000
; type <b> to boot
```

One thing about the Micropolis 1015 drives (the ones that pre-dated the later Tandom TM-100-4M drives) is that they have a head load signal (75ms load time) and very slow head step/settle times: 30ms step time, 10ms settle time.  If anything, I think this would help FlashFloppy, but I'm not sure.

Also check the drive interface signals on pp.3-5 of:
https://deramp.com/downloads/mfe_archive/010-S100%20Computers%20and%20Boards/01-Micropolis/30-Micropolis%20Disk%20Drives/Micropolis%201015-II%20Drive/102001A_Micropolis_1015_1016_Maintenance_Manual_Dec79.pdf

The drive select signals are a bit different than a PC floppy drive, and there is a head load signal which is on the same pin as a PC's density select signal.  Also, the 1015 drive provides a ready signal to the controller, and FlashFloppy may need to provide that, or you may be able to ground that signal at the controller side.

Take care,
Howard

**Reply by hharte** [reply]
@nullvalue0 Just saw your latest comment, great progress.  Now you just need a script to convert the .hfe back to .vgi to be able to verify.

I wonder how difficult it would to support .img images for Vector Graphic?  They are MFM, but the preamble and metadata are not an IBM format.  Also, even though CP/M only uses 256 bytes of the sector data, the sectors are 275 bytes each, and other Vector Graphic operating systems like MDOS and MZOS need the metadata.  This is why the .vgi file format exists.

**Reply by ejona86** [reply]
> the 1015 drive provides a ready signal to the controller

Yeah, and I see the the pinout on p.3-5 of Micropolis 1015/1016 maintenance manual matches that of the sheet 8 of the schematics (pdf page 47) in Micropolis Disk Controller Board. Pin 6 isn't connected on the Gotek, but grounding it does seem a fine approach.

This seems likely to be causing a problem.

> there is a head load signal which is on the same pin as a PC's density select signal

Looks like that can be ignored. It seems it isn't connected on the controller board.

> I wonder how difficult it would to support .img images for Vector Graphic?

I started with vgi in #401, but swapped to HFE later. The IMG code is completely different and there's no benefit to sharing. I think I mainly used fluxengine (w/ fluxengine or greaseweazle hardware) to read from flashfloppy and convert back to a IMG/VGI. (I have some local changes to fluxengine, but they conflict with @hharte's changes upstream; I also tend to get things ready and then David makes a big nice cleanup a day or two before I make a PR.)

(The Vector 4 uses ECC, which is nice and also wouldn't be in the 256 byte sector. I have finally reverse engineered it and have it working with Fluxengine, but just as a checksum/crc. I haven't gotten to actual error _correction_ yet. But this machine isn't that fancy.)

**Reply by nullvalue0** [reply]
@ejona86 @hharte Thanks for your continued help. 

Gave the FLOP2PC application a shot and it seems to have worked - read all the tracks off the gotek image I had loaded (VG09) and transferred the data over XMODEM to another computer. Very cool that it worked! Evaluating the data I ended up with....  It definitely has some of the same data that was in the original VG09.VGI file but it's nowhere near an exact match.. not sure if it'd help for me to upload it? 

So you think I should try tying pin 6 of the ribbon cable to GND at the controller end?

**Reply by nullvalue0** [reply]
> @ejona86: I started reading through https://bitsavers.org/pdf/vectorGraphic/hardware/Micropolis_Disk_Controller_Board.pdf just to find the boot disk number and the troubleshooting section was surprisingly unhelpful; I guess the "User's Information Sheet" (which I can't find)

BTW, I found that User's Information Sheet - it's in the document you linked to, it's just pages 6-7 in the PDF.

Taking a look at my FDC board, it appears the mods detailed on page 49 of that same PDF have already been performed. This updated the board from Rev 0 to Rev 1. One thing that this does essentially is ties pin 6 to GND.. so that is already done. I checked for continuity and it is in fact already tied to GND.
![PXL_20230113_155659297~3(1)](https://user-images.githubusercontent.com/58637612/212367058-0ccdde63-c522-4e1e-a3b7-d3f22b9084ca.jpg)

**Reply by hharte** [reply]
@nullvalue0 The Micropolis ROM bootstrap is described here on pp. 322 and a listing is provided: http://www.bitsavers.org/pdf/micropolis/metafloppy/1084-01_1040_1050_Users_Manual_Apr79.pdf

It looks like stepping in 8 tracks, and stepping out back to track zero is the first thing the bootstrap does after relocating itself to low RAM.

After that, the bootstrap tries to read track zero, sector zero into RAM.

Can you get back into the Vector ROM monitor with ESC when the boot attempt hangs?  If not, even resetting may be ok, and you could check the sector buffer in RAM to see what was actually read from the disk.

One possibility is that the wrong sector is being read, due to an issue with the index hole timing.  IIRC, there is a sector counter in HW that uses a one-shot to detect the index pulse, and reset the counter.  If you have a logic analyzer, you could probe that counter and see if it is counting and resetting properly.  There are 16 sector holes on the disk, and a 17th index hole that is half way between sector hole 15 and 0.

The Vector Monitor 4.0 is pretty decent, so you should be able to even set a breakpoint inside the bootstrap, although you’ll have to use the monitor to copy the relocatable part of the bootstrap to RAM, set the breakpoint in RAM, and then use the monitor to jump to the bootstrap in RAM.

It’s tedious, but you might be able to get an idea where things are going wrong.

Take care,
Howard

**Reply by nullvalue0** [reply]
@hharte Thank you, that explains why it's counting in tracks 0-7 then back out to 0. So I zeroed out all RAM and attempted to boot. It hung, but the memory is retained after a reset. It does in fact load some data between 0x009A and 0x015E. I took a checksum (Q 0000 1000) = 50. Checksum for the remainder of the memory is 00, so this is the only section that is modified. Then I tried booting a few other disks and it'd produce the exact same checksum and data in this block of memory (again, after zeroing each time). I thought that was a little odd but figure maybe the boot sector code is always the same. Then I tried booting a blank image and yet again I ended up with this same data. So.... not sure where this is coming from. I couldn't find this data anywhere in ROM, or in the original .VGI image. Any ideas?

I do have an 8-channel USB logic analyzer but I have never used to so may need a little more direction. I also have a full 2-channel digital scope if that'd be easier to use.

![PXL_20230115_192240527 MP~2](https://user-images.githubusercontent.com/58637612/212563165-a1ef9728-3e5e-40fe-b36b-dcb5c831ca75.jpg)

**Reply by hharte** [reply]
Hello @nullvalue0 ,

The code starting at 0A6H is the part of the bootstrap ROM that was copied to low RAM.  The intent of that piece of code is to read one sector into RAM.  You can see that the bootstrap ROM copied some of itself into RAM, and then the ROM code jumps to the label called SL010: (which is now in RAM.). That will select the drive, then do the step in/step out, and then jump back to the RDSEC: code at 0A6H to read the sector.

The sector data buffer is at 015Fh, and is still all 00’s, so presumably RDSEC: did not find what it’s looking for to start actually reading data from the controller into RAM.

see pp. 336 if the PDF linked above: “ WAIT FOR DESIRED SECTOR”

I imagine it’s reading the HW’s sector count register, waiting for it to get to zero, or it could be stuck waiting for the transfer flag.

With the machine in the state shown in your photo (boot loader copied to RAM) you could add a breakpoint at 00B6 and see if the bootstrap finds the sector count it’s looking for.

if it gets that far, then it might be stuck waiting for the transfer flag.  You could set a breakpoint at 00BB to see if it finds the transfer flag.

The HW read operation is described on pp.39 of http://www.s100computers.com/Hardware%20Manuals/Vector%20Graphic/VG%20Micropolis%20FDC%20Manual.pdf

They talk about the sector counter in Section 3.4 item D. You could hook up four channels of your logic analyzer to that counter’s S0-S3 outputs, and also monitor SECP (sector pulse) mentioned in item C.  You have three LA channels left, so you could hook them to the monostable multivibrator outputs mentioned in item E.  SECP should cause the counter to count from 0-15, and it should be reset by some combination of the SECP and multivibrator outputs.

It would also be good to monitor RDATA and RCLK mentioned in items G and H.  You could do another run capturing RDATA and RCLK along with SECP and the S0-S3 outputs.

References to the schematic locations of those signals are in the PDF.  Make sure to ground the LA to the board as well.

I’m not sure exactly which USB LA you have but those signals should be quite slow, so hopefully you can just stream a really long capture and not worry about setting up a trigger.  If you want to trigger on something, SECP is probably a good choice.

If you have a Micropolis drive and any 16-hard sector disk (even blank) you can capture the same signals with the real disk/drive and compare with FlashFloppy.

Take care,
Howard

**Reply by hharte** [reply]
Forgot to mention, after you set the breakpoint, “G 00D4” to start running from the bootstrap in RAM.

**Reply by nullvalue0** [reply]
@hharte Thank you for these explanations -- I'm starting to get a better picture of how this all works. I tried setting a breakpoint at 00B6 (and then again at 00BB for the heck of it) but I didn't get a register dump, so I'm guessing it didn't hit either breakpoint. This is the [logic analyzer](https://www.sparkfun.com/products/18627) I have so I think it should do the job.

**Reply by nullvalue0** [reply]
@hharte  Still running through some of these tests - I'm going through all of those read tests. Some things I'm seeing just don't match up to docs... For instance, 3.4.a (Test proper disk select) - I'm seeing all 4 lines start HIGH, then when I start the boot process I see DS1 AND DS4 go LOW. I would expect only one line to go low, correct?

Tests 3.4.b & c, I do see RDA and SECP pulses..

Test 3.4.d, I do see what looks like correct binary counting..

Then 3.4.e Check for proper pulse widths - I'm not seeing what looks like proper pulse widths. Looks like ~12.53ms pulse on most of those pins - but maybe I'm not checking the right pins. I'm going to run through all these tests one more time.

**Reply by nullvalue0** [reply]
@hharte Ok I repeated those tests and things are looking better. I must have had something hooked up wrong, the drive select is working correctly where only DS1 goes LOW at the start of the boot operation.

DS1 goes LOW, I see the SECP pulses. There is also activity on RDA, RDATA and RCLK.
![DS1, RDA, SECP, RDATA, RCLK](https://user-images.githubusercontent.com/58637612/213279426-581c4f7b-f63e-4577-8b52-8f23fb05abf3.png)

Next I tested the binary counting of S0-S3. This seems to be working fine from what I can tell?
![DS1, SECP, S0, S1, S2, S3](https://user-images.githubusercontent.com/58637612/213279722-074b2319-34a6-4faa-be9c-74a7eb47dc09.png)

Then I tested the monostable vibrators. I had to bump up the capture rate to 1MHz and there it was. I'm not sure if these are correct or what pins I should be testing exactly, but these are the results.
![DS1, SECP, MONO VIBR](https://user-images.githubusercontent.com/58637612/213280068-833c7b9f-c0f1-41ec-a779-0136e2ab4d94.png)

I have attached the [PulseView ](https://sigrok.org/wiki/Downloads) session output. If you want to be able to zoom in/pan on the data collected you could view the data in there. [Pulseview Session Data: pulseview.zip](https://github.com/keirf/flashfloppy/files/10450173/pulseview.zip). Each of the captures are only about 3 seconds long. They start just before the 'B' keypress from the VG Monitor, and end just after the Gotek displays the track count 0->7->0. The screenshots above may be slightly zoomed in to show a bit more detail and isn't quite the whole 3s capture. Please use PulseView to see the whole thing.

Let me know if this these tests tell you anything, or if there's another test I should carry out.

**Reply by ejona86** [reply]
Those traces look fine to me, from the flashfloppy perspective. I looked at them in PulseView and the sector pulses seem solid. RDATA needs over 2 MHz to see the data, as pulses are 400 ns, but gaps in RDATA due to stepping are still clearly visible. The cut-out is a steady 12 ms, which is the default `head-settle-ms`. Nothing looks out-of-the-ordinary.

**Reply by nullvalue0** [reply]
@ejona86 Thanks for reviewing - would it be helpful the see the RDATA line data?

**Reply by ejona86** [reply]
It'd allow seeing if anything is wrong with the emulator, although the situation here looks pretty simple. RDATA and SECP at 3+ MHz would be great.

**Reply by nullvalue0** [reply]
I have captured RDATA @ 4MHz - probably can't tell much by the screenshot, will have to use PulseView again - session attached below.

![DS1, SECP, RDATA, RCLK @ 4MHz](https://user-images.githubusercontent.com/58637612/213816927-ae35ba1e-fc08-4fc7-a0d8-549f0f868150.png)

[DS1, SECP, RDATA, RCLK @ 4MHz.zip](https://github.com/keirf/flashfloppy/files/10470675/DS1.SECP.RDATA.RCLK.%40.4MHz.zip)

**Reply by ejona86** [reply]
@nullvalue0, the data in that capture looks perfect. I don't see any hiccups. The index pulses look great and the alignment with the data looks great, even when seeking. I decoded the data and the checksums were valid. A spot-check of the first sector showed the same data as VG09.

**Reply by nullvalue0** [reply]
Wow.. Thanks so much for reviewing, that's good news..... I think. That means I must have a hardware problem somewhere. There were a few more Read Operation tests in the page that @hharte linked to, I'll try carrying those out.

**Reply by nullvalue0** [reply]
@ejona86 @hharte So the final tests it talks about is checking the DIG signal. It's definitely doing something when the disk read operation starts, then once it reaches the part where it hangs, the DIG pin continues pulsing.
![DS1, SECP, RCLK, RDATA, DIG](https://user-images.githubusercontent.com/58637612/213900623-aa63b4a8-9301-4686-a117-9d064648079d.png)

Then it says to check for proper operation of the output data buffers at D11 and D12. I did these in 2 separate tests. Both chips' results look similar. It's basically just sending data all the time it seems - regardless of whether the disk is being read or not. I'm not sure if this is expected behavior? The PulseView results are also uploaded below.
![DS1, SECP, RCLK, RDATA, D11](https://user-images.githubusercontent.com/58637612/213900777-c870daf8-299f-44fc-a39c-1e5108e1c432.png)
![DS1, SECP, RCLK, RDATA, D12](https://user-images.githubusercontent.com/58637612/213900776-aa1bef54-bc6e-4fc7-8bfa-2a3891779bf2.png)


[PulseView3.zip](https://github.com/keirf/flashfloppy/files/10473747/PulseView3.zip)

**Reply by nullvalue0** [reply]
Do you think I'd have better luck if I tried switching to a different FDC like the [Dual Mode FD HD](http://www.s100computers.com/Hardware%20Folder/Vector%20Graphic/Dual%20FD-HD%20Controller/VG%20Dual%20HDC.htm) board? @ejona86 which controller do you have in your Vector 4?

**Reply by ejona86** [reply]
If you have a Dual-Mode board and it would work with your system, that is what I use and it is solid. The same PCB was used for an FD-only version of it that also works well. I'm pretty confident every Vector 4 used one of those two boards, and I've only used them in the Vector 4. Be aware the Vector 4 used a modified S-100 bus with regulated power lines.

You can find [its manual and schematic on bitsavers](https://bitsavers.org/pdf/vectorGraphic/hardware/7200-1200-02-1_Dual-Mode_Disk_Controller_Board_Engineering_Documentation_Feb81.pdf). s100computers also a schematic, but I've been using the one at the end of the [Vector 4 Technical Information](https://archive.org/details/7200-0001-vector-4-technical-information-sep-82).
